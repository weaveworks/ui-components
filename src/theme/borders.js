import { kebabCase, forEach } from 'lodash';

export const borderRadius = {
  circle: '50%',
  none: '0',
  soft: '2px',
};

// Collects all theme border radii as SCSS vars.
export function themeBorderRadiiAsScss() {
  const themeBorderRadii = [];

  forEach(borderRadius, (value, name) => {
    themeBorderRadii.push(`$border-radius-${kebabCase(name)}: ${value}`);
  });

  return themeBorderRadii;
}
