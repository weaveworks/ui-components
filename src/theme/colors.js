import { isString, kebabCase, forEach } from 'lodash';

export const colors = {
  // Neutral Colors
  black: 'hsl(0, 0%, 10%)', // #1a1a1a
  blue50: 'hsl(191, 100%, 97%)', // #f0fcff
  blue200: 'hsl(191, 100%, 80%)', // #99ecff
  blue400: 'hsl(191, 100%, 50%)', // #00d2ff
  blue600: 'hsl(191, 100%, 40%)', // #00a7cc
  blue700: 'hsl(191, 100%, 35%)', // #0092b3
  blue800: 'hsl(191, 100%, 30%)', // #007d99
  graphThemes: {
    blue: [
      '#c7e9b4',
      '#7ecdbb',
      '#1eb5eb',
      '#1d91bf',
      '#235fa9',
      '#253393',
      '#084181',
    ],
    mixed: [
      '#c7e9b4',
      '#c1d4e7',
      '#7ecdbb',
      '#9fbddb',
      '#1eb5eb',
      '#8d95c6',
      '#1d91bf',
      '#8282ab',
      '#235fa9',
      '#89429e',
      '#253393',
      '#800f7a',
      '#084181',
      '#0b0533',
    ],
    purple: [
      '#c1d4e7',
      '#9fbddb',
      '#8d95c6',
      '#8282ab',
      '#89429e',
      '#800f7a',
      '#0b0533',
    ],
  }, // Used by PrometheusGraph component
  gray50: 'hsl(0, 0%, 96%)', // #f4f4f4
  gray100: 'hsl(0, 0%, 90%)', // #e6e6e6
  gray200: 'hsl(0, 0%, 80%)', // #cccccc

  // #737373
  gray600: 'hsl(0, 0%, 45%)',
  green500: 'hsl(161, 54%, 48%)', // #38bd93
  orange500: 'hsl(13, 100%, 50%)', // #ff3700
  orange600: 'hsl(13, 100%, 40%)', // #cc2c00
  orange700: 'hsl(13, 100%, 35%)', // #b32700
  // Accent Colors
  // #992100
  orange800: 'hsl(13, 100%, 30%)',
  promQL: {
    attrName: '#00a4db',
    /**
     * GHColors theme by Avi Aryan (http://aviaryan.in)
     * Inspired by Github syntax coloring
     */
    comment: '#bbbbbb',
    deleted: '#9a050f',
    entity: '#36acaa',
    function: '#dc322f',
    metricName: '#2aa198',
    punctuation: '#393a34',
    // Dropdown colors,
    salmon: '#ff7c7c',
    string: '#e3116c',
    tag: '#00009f',
  }, // PromQL
  purple25: 'hsl(240, 20%, 98%)', // #fafafc
  purple50: 'hsl(240, 20%, 95%)', // #eeeef4
  purple100: 'hsl(240, 20%, 90%)', // #dfdfea
  purple200: 'hsl(240, 20%, 75%)', // #b1b1cb
  purple300: 'hsl(240, 20%, 65%)', // #9494b8
  purple400: 'hsl(240, 20%, 60%)', // #8585ad

  purple500: 'hsl(240, 20%, 50%)', // #666699
  purple600: 'hsl(240, 20%, 45%)', // #5b5b88
  purple700: 'hsl(240, 20%, 35%)', // #47476b
  purple800: 'hsl(240, 20%, 30%)', // #3d3d5c
  // Primary Colors
  // #32324b
  purple900: 'hsl(240, 20%, 25%)',

  // Third-party specific colors - not to be used in the theme!
  thirdParty: {
    azure: '#3769bb',
    // Google single-click login
    cornflowerBlue: '#4285f4',
  },
  // #ffffff
  white: 'hsl(0, 0%, 100%)',
  // #d4ab27
  yellow500: 'hsl(46, 69%, 49%)',
};

// Flattens and collects all theme colors as SCSS vars
export function themeColorsAsScss() {
  const ignoreKeys = ['graphThemes'];
  const themeColors = [];

  forEach(colors, (value, name) => {
    const colorPrefix = `$color-${kebabCase(name)}`;
    if (ignoreKeys.includes(name)) return;

    if (isString(value)) {
      themeColors.push(`${colorPrefix}: ${value}`);
    } else {
      forEach(value, (innerValue, subname) => {
        if (ignoreKeys.includes(subname)) return;
        themeColors.push(`${colorPrefix}-${kebabCase(subname)}: ${innerValue}`);
      });
    }
  });

  return themeColors;
}
