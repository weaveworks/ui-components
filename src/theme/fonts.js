import { kebabCase, forEach } from 'lodash';

export const fontFamilies = {
  monospace: "'Roboto Mono', monospace",
  regular: "'proxima-nova', Helvetica, Arial, sans-serif",
};

export const fontSizes = {
  extraLarge: '32px',
  huge: '48px',
  large: '22px',
  normal: '16px',
  small: '14px',
  tiny: '12px',
};

// Collects all theme font vars as SCSS vars.
export function themeFontsAsScss() {
  const themeFonts = [];

  forEach(fontFamilies, (value, name) => {
    themeFonts.push(`$font-family-${kebabCase(name)}: ${value}`);
  });

  forEach(fontSizes, (value, name) => {
    themeFonts.push(`$font-size-${kebabCase(name)}: ${value}`);
  });

  return themeFonts;
}
