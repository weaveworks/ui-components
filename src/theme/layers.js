import { kebabCase, forEach } from 'lodash';

export const layers = {
  alert: 3,
  dropdown: 5,
  front: 1,
  modal: 7,
  notification: 4,
  toolbar: 2,
  tooltip: 6,
};

// Collects all theme z-index layers as SCSS vars
export function themeLayersAsScss() {
  const themeLayers = [];

  forEach(layers, (value, name) => {
    themeLayers.push(`$layer-${kebabCase(name)}: ${value}`);
  });

  return themeLayers;
}
