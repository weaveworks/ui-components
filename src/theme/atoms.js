import { colors } from './colors';

// component-specific
export const atoms = {
  Alert: {
    error: {
      background: colors.orange600,
      color: colors.white,
    },
    info: {
      background: colors.blue600,
      color: colors.white,
    },
    success: {
      background: colors.green500,
      color: colors.white,
    },
    warning: {
      background: colors.yellow500,
      color: colors.white,
    },
  },
  Button: {
    danger: {
      background: colors.orange600,
      color: colors.white,
      hoverBackground: colors.orange700,
      hoverColor: colors.white,
    },
    default: {
      background: colors.white,
      color: colors.black,
      hoverBackground: colors.gray50,
      hoverColor: colors.purple800,
    },
    disabled: {
      background: colors.gray50,
      color: colors.gray600,
      hoverBackground: colors.gray50,
      hoverColor: colors.gray600,
    },
    primary: {
      background: colors.blue700,
      color: colors.white,
      hoverBackground: colors.blue800,
      hoverColor: colors.white,
    },
  },
};
