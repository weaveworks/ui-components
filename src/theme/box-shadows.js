import { colors } from './colors';

export const boxShadow = {
  heavy: '0 10px 20px rgba(0,0,0,0.19), 0 6px 6px rgba(0,0,0,0.23)',
  light: '0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24)',
  none: 'none',
  selected: `0px 0px 2px 2px ${colors.purple400}`,
};
