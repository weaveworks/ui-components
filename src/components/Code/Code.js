import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';
import { isFunction, isString, trim, noop } from 'lodash';

import { spacing } from '../../theme/selectors';

const scale = keyframes`
  0% {
    transform: scale(0)
  }
  90% {
    transform: scale(1.2)
  }
  100% {
    transform: scale(1)
  }
`;

const CopyNotice = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  padding: ${spacing('xs')};
  ${props => props.isHovered && 'transition: opacity 300ms ease;'};
  border-radius: ${props => props.theme.borderRadius.soft};
  background-color: ${props => props.theme.colors.purple800};
  opacity: 0;
  font-size: ${props => props.theme.fontSizes.small};
  color: ${props => props.theme.colors.purple50};
  ${props =>
    props.isCopying &&
    `
    & > i {
        transform-origin: center;
        animation-name: ${scale};
        animation-duration: .4s;
      }
  `};
`;

const CodeWrapper = styled.div`
  position: relative;
  background-color: ${props => props.theme.colors.purple800};
  box-sizing: border-box;
  border-radius: ${props => props.theme.borderRadius.soft};
  cursor: pointer;

  &:hover ${CopyNotice} {
    opacity: 0.9;
  }
`;

const ScrollWrap = styled.div`
  overflow: auto;
`;

const Content = styled.div`
  flex-grow: 1;
  color: ${props => props.theme.colors.purple50};
  padding: ${spacing('xs')} ${spacing('base')};
`;

const Pre = styled.pre`
  margin: ${spacing('small')} 0;
  font-family: ${props => props.theme.fontFamilies.monospace};
  font-size: ${props => props.theme.fontSizes.small};
`;

const PaddedLine = styled.div`
  &:not(:last-child) {
    padding-bottom: ${spacing('xs')};
  }

  /* required to make jsx children work with adding '\n' in multiLine */
  & > * {
    display: inline;
  }
`;

const trimString = node => (isString(node) ? trim(node) : node);

const formatSingleCommand = children =>
  isFunction(children) ? children() : trimString(children);

function formatMultiString(string) {
  return trim(string)
    .split('\n')
    .map(line => (
      <PaddedLine key={line}>
        {line}
        {'\n'}
      </PaddedLine>
    ));
}

function formatMultiCommand(raw) {
  let children = raw;

  if (isFunction(children)) {
    children = children();
  }
  const count = React.Children.count(children);

  if (count === 1) {
    return isString(children) ? formatMultiString(children) : children;
  }

  return React.Children.map(children, (child, i) => (
    <PaddedLine key={i}>
      {child}
      {'\n'}
    </PaddedLine>
  ));
}

/**
 * Code allows for easy rendering of code snippets which can easliy be copied to
 * the clipboard by clicking the element or selecting a portion of the code
 */
class Code extends Component {
  static displayName = 'Code';

  state = {
    isCopying: false,
    isHovered: false,
  };

  componentDidUpdate = () => {
    // reselect text after re-render
    if (this.state.selectedRange) {
      const prevRange = this.state.selectedRange;
      const sel = window.getSelection();
      sel.removeAllRanges();
      sel.addRange(prevRange);
      // clear it for the next interaction
      this.setState({ selectedRange: null });
    }
  };

  handleCopyClick = () => {
    // Create a textarea element in which the text of the code will be copied.
    // Set the value of the textarea, select it to set it as the active element,
    // then copy the value of that element to the clipboard.
    // It's 2017 and this is still the best way to copy text on button click...
    if (isFunction(this.props.onCopy)) {
      this.props.onCopy();
    }
    const selection = document.getSelection();
    const selectionString = selection.toString();
    // if user has selected text, save that to state so it can be reselected
    if (selectionString !== '') {
      this.setState({
        selectedRange: selection.getRangeAt(0),
      });
    }

    const code =
      selectionString === '' ? this.preNode.textContent : selectionString;
    const txtArea = document.createElement('textarea');
    // Safari doesn't allow for assigning an object literal to `style`.
    txtArea.class = 'hidden-textarea';
    // Make sure the code ends in a newline to execute all the commands in multiline code.
    if (code.charAt(code.length - 1) === '\n') {
      txtArea.value = code;
    } else {
      txtArea.value = `${code}\n`;
    }
    document.body.appendChild(txtArea);

    txtArea.select();

    try {
      document.execCommand('copy');
    } catch (e) {
      throw e;
    }
    document.body.removeChild(txtArea);

    // show the Copied to clipboard notice temporarily
    this.setState({ isCopying: true });
    setTimeout(() => {
      this.setState({ isCopying: false });
    }, 3000);
  };

  onMouseEnter = () => {
    this.setState({
      isHovered: true,
    });
  };

  onMouseLeave = () => {
    this.setState({
      isCopying: false,
      isHovered: false,
    });
  };

  render() {
    const { children, multiCommand } = this.props;
    const { isCopying, isHovered } = this.state;

    const copy =
      isCopying && isHovered ? (
        <i className="fa fa-check" />
      ) : (
        'Copy to clipboard'
      );

    return (
      <CodeWrapper
        onClick={this.handleCopyClick}
        onMouseEnter={this.onMouseEnter}
        onMouseLeave={this.onMouseLeave}
      >
        <ScrollWrap>
          <Content>
            <Pre
              innerRef={e => {
                this.preNode = e;
              }}
            >
              {multiCommand
                ? formatMultiCommand(children)
                : formatSingleCommand(children)}
            </Pre>
          </Content>
        </ScrollWrap>

        <CopyNotice isCopying={isCopying} isHovered={isHovered}>
          {copy}
        </CopyNotice>
      </CodeWrapper>
    );
  }
}

Code.propTypes = {
  // children can be anything that React can render
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,

  // multiCommand determines if the code lines shall be padded
  multiCommand: PropTypes.bool,

  // onCopy will be called when the CodeWrapper is clicked
  onCopy: PropTypes.func,
};

Code.defaultProps = {
  multiCommand: false,
  onCopy: noop,
};

export default Code;
