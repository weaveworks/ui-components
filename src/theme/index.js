import { colors, themeColorsAsScss } from './colors';
import { fontSizes, fontFamilies, themeFontsAsScss } from './fonts';
import { borderRadius, themeBorderRadiiAsScss } from './borders';
import { layers, themeLayersAsScss } from './layers';
import { spacing, themeSpacingsAsScss } from './spacings';
import { overlayIconSize, textColor, themeMiscVarsAsScss } from './misc';
import { boxShadow } from './box-shadows';
import { atoms } from './atoms';

export function themeVarsAsScss() {
  const themeVariables = []
    .concat(themeColorsAsScss())
    .concat(themeLayersAsScss())
    .concat(themeFontsAsScss())
    .concat(themeBorderRadiiAsScss())
    .concat(themeSpacingsAsScss())
    .concat(themeMiscVarsAsScss());
  return `${themeVariables.join('; ')};`;
}

const theme = {
  // Component-specific
  atoms,

  // Borders
  borderRadius,

  // Box shadows
  boxShadow,

  // Colors
  colors,

  // Fonts
  fontFamilies,
  fontSizes,

  // z-index layers
  layers,

  // Misc
  overlayIconSize,

  // Spacings
  spacing,
  textColor,
};

export default theme;
