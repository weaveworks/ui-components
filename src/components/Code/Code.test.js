import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { withTheme } from '../../utils/theme';

import { CopyNotice } from './Code';

import Code from '.';

const getSnapShot = cmp => renderer.create(withTheme(cmp)).toJSON();

document.execCommand = () => {};
document.getSelection = () => ({
  toString: () => '',
});

describe('<Code />', () => {
  it('should call onCopy when the button is clicked', () => {
    const onCopy = jest.fn();
    const wrapper = mount(
      withTheme(<Code onCopy={onCopy}>such code much programming</Code>)
    );
    expect(onCopy).not.toHaveBeenCalled();
    wrapper.find(CopyNotice).simulate('click');
    expect(onCopy).toHaveBeenCalledTimes(1);
  });

  it('should render correctly', () => {
    const snapshot = getSnapShot(<Code>such code much programming</Code>);
    expect(snapshot).toMatchSnapshot();
  });

  it('should render multi line string correctly with multiCommand=true', () => {
    const snapshot = getSnapShot(
      <Code multiCommand>
        {`
sudo curl -l git.io/scope -o /usr/local/bin/scope
sudo chmod a+x /usr/local/bin/scope
scope launch https://yjsjsubdx1h8un1f858gp7to8d51zdre@frontend.dev.weave.works
          `}
      </Code>
    );
    expect(snapshot).toMatchSnapshot();
  });

  it('should render multi line function correctly with multiCommand=true', () => {
    const snapshot = getSnapShot(
      <Code multiCommand>
        {() => `
sudo curl -l git.io/scope -o /usr/local/bin/scope
sudo chmod a+x /usr/local/bin/scope
scope launch https://yjsjsubdx1h8un1f858gp7to8d51zdre@frontend.dev.weave.works
      `}
      </Code>
    );
    expect(snapshot).toMatchSnapshot();
  });

  it('should render multi line jsx correctly with multiCommand=true', () => {
    const snapshot = getSnapShot(
      <Code multiCommand>
        <div>sudo curl -L git.io/scope -o /usr/local/bin/scope</div>
        <div>sudo chmod a+x /usr/local/bin/scope</div>
        <div>
          scope launch <wbr />
          https://yjsjsubdx1h8un1f858gp7to8d51zdre@frontend.dev.weave.works
        </div>
      </Code>
    );
    expect(snapshot).toMatchSnapshot();
  });
});
