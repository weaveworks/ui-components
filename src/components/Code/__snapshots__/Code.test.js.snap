// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Code /> should render correctly 1`] = `
<div
  className="Code__CodeWrapper-sc-1no9z1g-1 jTyFrG"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="Code__ScrollWrap-sc-1no9z1g-2 djYguF"
  >
    <div
      className="Code__Content-sc-1no9z1g-3 dsHiaX"
    >
      <pre
        className="Code__Pre-sc-1no9z1g-4 bxgykZ"
      >
        such code much programming
      </pre>
    </div>
  </div>
  <div
    className="Code__CopyNotice-sc-1no9z1g-0 cuyWgG"
  >
    Copy to clipboard
  </div>
</div>
`;

exports[`<Code /> should render multi line function correctly with multiCommand=true 1`] = `
<div
  className="Code__CodeWrapper-sc-1no9z1g-1 jTyFrG"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="Code__ScrollWrap-sc-1no9z1g-2 djYguF"
  >
    <div
      className="Code__Content-sc-1no9z1g-3 dsHiaX"
    >
      <pre
        className="Code__Pre-sc-1no9z1g-4 bxgykZ"
      >
        <div
          className="Code__PaddedLine-sc-1no9z1g-5 cWhzHF"
        >
          sudo curl -l git.io/scope -o /usr/local/bin/scope
          

        </div>
        <div
          className="Code__PaddedLine-sc-1no9z1g-5 cWhzHF"
        >
          sudo chmod a+x /usr/local/bin/scope
          

        </div>
        <div
          className="Code__PaddedLine-sc-1no9z1g-5 cWhzHF"
        >
          scope launch https://yjsjsubdx1h8un1f858gp7to8d51zdre@frontend.dev.weave.works
          

        </div>
      </pre>
    </div>
  </div>
  <div
    className="Code__CopyNotice-sc-1no9z1g-0 cuyWgG"
  >
    Copy to clipboard
  </div>
</div>
`;

exports[`<Code /> should render multi line jsx correctly with multiCommand=true 1`] = `
<div
  className="Code__CodeWrapper-sc-1no9z1g-1 jTyFrG"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="Code__ScrollWrap-sc-1no9z1g-2 djYguF"
  >
    <div
      className="Code__Content-sc-1no9z1g-3 dsHiaX"
    >
      <pre
        className="Code__Pre-sc-1no9z1g-4 bxgykZ"
      >
        <div
          className="Code__PaddedLine-sc-1no9z1g-5 cWhzHF"
        >
          <div>
            sudo curl -L git.io/scope -o /usr/local/bin/scope
          </div>
          

        </div>
        <div
          className="Code__PaddedLine-sc-1no9z1g-5 cWhzHF"
        >
          <div>
            sudo chmod a+x /usr/local/bin/scope
          </div>
          

        </div>
        <div
          className="Code__PaddedLine-sc-1no9z1g-5 cWhzHF"
        >
          <div>
            scope launch 
            <wbr />
            https://yjsjsubdx1h8un1f858gp7to8d51zdre@frontend.dev.weave.works
          </div>
          

        </div>
      </pre>
    </div>
  </div>
  <div
    className="Code__CopyNotice-sc-1no9z1g-0 cuyWgG"
  >
    Copy to clipboard
  </div>
</div>
`;

exports[`<Code /> should render multi line string correctly with multiCommand=true 1`] = `
<div
  className="Code__CodeWrapper-sc-1no9z1g-1 jTyFrG"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="Code__ScrollWrap-sc-1no9z1g-2 djYguF"
  >
    <div
      className="Code__Content-sc-1no9z1g-3 dsHiaX"
    >
      <pre
        className="Code__Pre-sc-1no9z1g-4 bxgykZ"
      >
        <div
          className="Code__PaddedLine-sc-1no9z1g-5 cWhzHF"
        >
          sudo curl -l git.io/scope -o /usr/local/bin/scope
          

        </div>
        <div
          className="Code__PaddedLine-sc-1no9z1g-5 cWhzHF"
        >
          sudo chmod a+x /usr/local/bin/scope
          

        </div>
        <div
          className="Code__PaddedLine-sc-1no9z1g-5 cWhzHF"
        >
          scope launch https://yjsjsubdx1h8un1f858gp7to8d51zdre@frontend.dev.weave.works
          

        </div>
      </pre>
    </div>
  </div>
  <div
    className="Code__CopyNotice-sc-1no9z1g-0 cuyWgG"
  >
    Copy to clipboard
  </div>
</div>
`;
